p2p/proto/consensus.proto

```diff
@@ -1,42 +1,80 @@
 syntax = "proto3";
 
-import "p2p/proto/common.proto";
-import "p2p/proto/transaction.proto";
-import "p2p/proto/state.proto";
+import "common.proto";
+import "transaction.proto";
+import "block.proto";
 import "google/protobuf/timestamp.proto";
 
 
 // WIP - will change
 
-message Proposal {
-    uint64                    block_number      = 2;
-    uint32                    round             = 3;
-    uint32                    pol               = 4;  // proof of lock
-    Hash                      block_header_hash = 5;
-    google.protobuf.Timestamp timestamp         = 6;
-    ConsensusSignature        signature         = 7;
+
+message BlockLocation {
+    Hash   parent_hash = 1;
+    uint64 height      = 2;
+    uint64 fork_id     = 3;
 }
 
-// A block proposal is a series of (Transactions+, StateDiff)* BlockHeader
+enum Step {
+    Propose = 1;
+    Prevote = 2;
+    Precommit = 3;
+}
 
 message Vote {
-    enum  Type {
-        UNKNOWN   = 0;
-        Proposal  = 1;
-        Prevote   = 2;
-        Precommit = 3;
+    enum  VoteType {
+        Prevote   = 1;
+        Precommit = 2;
     };
 
-    Proposal           proposal          = 1;
-    Address            validator_address = 2;
-    int32              validator_index   = 3;  // ???
-    ConsensusSignature signature         = 4;
+    // We use a type field to distinguish between prevotes and precommits instead of different
+    // messages, to make sure the data, and therefore the signatures, are unambiguous between
+    // Prevote and Precommit.
+    VoteType           vote_type      = 3;
+    BlockLocation      block_location = 4;
+    uint32             round          = 5;
+    // This is optional since a vote can be NIL.
+    optional Hash      block_hash     = 6;
+    // Identifies the voter.
+    Address            voter          = 7;
+    // Voter's signature of the above fields.
+    ConsensusSignature signature      = 8;
+}
+
+message ProposalInit {
```

Comment:
Note that after unifying proposal and vote messages, proposal init looks a lot like a vote.

Reply:
They certainly have some similar fields, but I fear that it would just make the protobuf source code harder to understand if we used the same struct. As in ProposalInit becomes Vote but with `step=Propose` (instead of VoteType) and `block_hash=None`, it seems weird to call the struct `Vote`. But yes we could do this if you feel strongly.

We'd also have some Vote wrapper to move the signature there since ProposalInit doesn't itself have signature.

Reply:
But the Proposal message does. Anyway, I don't feel strongly about it.